File: examples/ImpactMetricExample.py
#!/usr/bin/env python
#==========================================================================
#
#   Copyright NumFOCUS
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#          https://www.apache.org/licenses/LICENSE-2.0.txt
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.
#
#==========================================================================

"""IMPACT similarity metric: a feature-based ImageToImageMetricv4 for registration.

``itk.ImpactImageToImageMetricv4`` compares two images through TorchScript features
instead of raw intensities, so it plugs into ``itk.ImageRegistrationMethodv4`` and drives
any ITK transform/optimizer -- while working across modalities. Passing the modality
invariant MIND descriptor with a normalized-cross-correlation distance gives a smooth,
well-behaved metric for e.g. CT-to-MR alignment (no ``import torch`` needed).

This example recovers a rigid misalignment between a CT (fixed) and an MR (moving).

Run:  ./ImpactMetricExample.py fixed_ct.mha moving_mr.mha mind.pt \\
          --out moved.mha --device cuda
"""

import argparse
import itk

parser = argparse.ArgumentParser(description="IMPACT metric v4 rigid registration demo.")
parser.add_argument("fixed_image")
parser.add_argument("moving_image")
parser.add_argument("models", nargs="+", help="TorchScript feature model(s), e.g. a MIND descriptor")
parser.add_argument("--out", default="moved.mha")
parser.add_argument("--iterations", type=int, default=60)
parser.add_argument("--voxel", type=float, default=2.0)
parser.add_argument("--mask", default=None,
                    help="body mask on the fixed image; strongly recommended (see below)")
parser.add_argument("--device", default="cuda", help='"cpu", "cuda", "cuda:0", ...')
args = parser.parse_args()

Dimension = 3
ImageType = itk.Image[itk.F, Dimension]

fixed = itk.imread(args.fixed_image, itk.F)
moving = itk.imread(args.moving_image, itk.F)

voxel = [args.voxel] * Dimension

metric = itk.ImpactImageToImageMetricv4[ImageType, ImageType].New()
# Add one configuration per model (applies to both fixed and moving images).
for path in args.models:
    metric.AddModelConfiguration(
        itk.ImpactModelConfiguration(path, Dimension, 1, [0, 0, 0], voxel, 0, [True], False)
    )
metric.SetDistance(["L2"] * len(args.models))      # the distance to reach for; NCC is an ablation
metric.SetLayersWeight([1.0] * len(args.models))
metric.SetSubsetFeatures([12] * len(args.models))  # channels compared per model
metric.SetPCA([0] * len(args.models))
metric.SetMode("Static")
metric.SetDevice(args.device)
if args.mask:
    # Air agrees between modalities everywhere, so a whole-image domain measures mostly
    # background: on an abdominal MR-CT pair the similarity varies by 1e-4 over a 2 cm offset,
    # which no optimizer can follow, against 5e-2 for the same offset inside the body.
    #
    # How tight the mask is decides the result more than any other setting here. A whole-body
    # mask is mostly wall and fat, which agree well between the two modalities and outvote the
    # organs: on the pair below it costs 0.14 of organ Dice against a mask around the organs
    # themselves, and no distance, model or optimizer setting recovers that. A mask of the
    # region you want aligned is worth more than a better similarity.
    mask = itk.ImageMaskSpatialObject[Dimension].New()
    mask.SetImage(itk.imread(args.mask, itk.UC))
    mask.Update()
    metric.SetFixedImageMask(mask)

# Rigid transform, centred on the fixed image, which is the domain the moving transform maps
# from. The centre goes through the direction cosines: computing it as origin + 0.5 * spacing *
# size ignores them, and on an image stored LPS rather than RAS that lands hundreds of
# millimetres outside the anatomy. Every degree of rotation then arrives with a lever arm, and
# the registration settles on a transform that translates the organs instead of turning them.
size = fixed.GetLargestPossibleRegion().GetSize()
center = list(fixed.TransformIndexToPhysicalPoint([size[0] // 2, size[1] // 2, size[2] // 2]))
transform = itk.Euler3DTransform[itk.D].New()
transform.SetIdentity()
transform.SetCenter(center)

optimizer = itk.RegularStepGradientDescentOptimizerv4[itk.D].New()
optimizer.SetLearningRate(2.0)
optimizer.SetMinimumStepLength(1e-4)
optimizer.SetNumberOfIterations(args.iterations)
optimizer.SetRelaxationFactor(0.8)
# Let ITK derive the parameter scales from the physical shift each parameter causes, rather
# than guessing a rotation-versus-translation ratio by hand: the right ratio depends on the
# metric's own gradient magnitude, and a wrong guess walks the transform off the image.
estimator = itk.RegistrationParameterScalesFromPhysicalShift[type(metric)].New()
estimator.SetMetric(metric)
optimizer.SetScalesEstimator(estimator)

registration = itk.ImageRegistrationMethodv4[ImageType, ImageType].New()
registration.SetFixedImage(fixed)
registration.SetMovingImage(moving)
registration.SetMetric(metric)
registration.SetOptimizer(optimizer)
registration.SetInitialTransform(transform)
registration.SetNumberOfLevels(1)
registration.SetSmoothingSigmasPerLevel([0])
registration.SetShrinkFactorsPerLevel([1])
registration.Update()

final = registration.GetTransform()
print("recovered parameters:", [round(v, 3) for v in final.GetParameters()])

moved = itk.resample_image_filter(
    moving, transform=final, use_reference_image=True, reference_image=fixed, default_pixel_value=0.0
)
itk.imwrite(moved, args.out)
print("wrote", args.out)

File: examples/ImpactMetricExample.cxx
/*=========================================================================
 *
 *  Copyright NumFOCUS
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *         https://www.apache.org/licenses/LICENSE-2.0.txt
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *=========================================================================*/

// IMPACT: semantic similarity registration from pretrained TorchScript features.
//
// This example mirrors ImpactMetricExample.py and shows the two ITK-facing pieces of
// the module. No torch type ever crosses the API: it speaks only ITK/STL types (images,
// a model-path string, a device string).
//
//   1. the core:   itk::ImageToFeaturesMap extracts a feature map from an image;
//   2. the metric: itk::ImpactImageToImageMetricv4 plugs into
//                  itk::ImageRegistrationMethodv4 to register two images by comparing
//                  those features instead of raw intensities.

#include "itkImageToFeaturesMap.h"
#include "itkImpactImageToImageMetricv4.h"
#include "itkModelConfiguration.h"

#include <itkImageFileReader.h>
#include <itkImageFileWriter.h>
#include <itkImageRegistrationMethodv4.h>
#include <itkRegularStepGradientDescentOptimizerv4.h>
#include <itkTranslationTransform.h>
#include <itkResampleImageFilter.h>
#include <itkBSplineInterpolateImageFunction.h>

int
main(int argc, char * argv[])
{
  if (argc < 5)
  {
    std::cerr << "IMPACT: semantic similarity registration from pretrained features.\n";
    std::cerr << "Usage: " << argv[0] << " model.pt fixedImage movingImage outputWarpedImage [device]\n";
    std::cerr << "  device: \"cpu\" (default), \"cuda\", \"cuda:0\", ...\n";
    return EXIT_FAILURE;
  }
  const char * const  modelPath = argv[1];
  const char * const  fixedPath = argv[2];
  const char * const  movingPath = argv[3];
  const char * const  outputPath = argv[4];
  const std::string   device = (argc > 5) ? argv[5] : "cpu";

  constexpr unsigned int Dimension = 3;
  using PixelType = float;
  using ImageType = itk::Image<PixelType, Dimension>;

  const auto fixed = itk::ReadImage<ImageType>(fixedPath);
  const auto moving = itk::ReadImage<ImageType>(movingPath);

  // A TorchScript model configuration: (path, dimension, channels, patchSize, voxelSize,
  // overlap, layersMask, mixedPrecision). Only POD/STL types cross the API.
  const itk::ModelConfiguration config(
    modelPath, Dimension, 1, { 0, 0, 0 }, { 1.0f, 1.0f, 1.0f }, 2, { true }, false);

  // --- 1. core: extract a dense feature map from the fixed image --------------------
  using InterpolatorType = itk::BSplineInterpolateImageFunction<ImageType, double>;
  auto interpolator = InterpolatorType::New();
  interpolator->SetSplineOrder(3);

  auto features = itk::ImageToFeaturesMap<ImageType, InterpolatorType>::New();
  features->SetModelConfiguration(config);
  features->SetInterpolator(interpolator);
  features->SetDevice(device);
  features->AddInput(fixed);
  features->Update();
  const auto featureMap = features->GetOutput(0); // itk::VectorImage<float, 3>
  std::cout << "feature map: " << featureMap->GetLargestPossibleRegion().GetSize()
            << "  channels: " << featureMap->GetNumberOfComponentsPerPixel() << std::endl;

  // --- 2. metric: register moving onto fixed by comparing anatomical features -------
  using MetricType = itk::ImpactImageToImageMetricv4<ImageType, ImageType>;
  auto                                 metric = MetricType::New();
  std::vector<itk::ModelConfiguration> models{ config };
  metric->SetModelsConfiguration(models);
  metric->SetDistance({ "NCC" });      // per-layer loss: L1, L2, NCC, Cosine, Dice, ...
  metric->SetLayersWeight({ 1.0f });
  metric->SetSubsetFeatures({ 4 });    // random channel subset for speed (0 = all)
  metric->SetPCA({ 0 });
  metric->SetMode("Static");           // "Static" (precomputed features) or "Jacobian"
  metric->SetDevice(device);

  using TransformType = itk::TranslationTransform<double, Dimension>;
  auto transform = TransformType::New();
  transform->SetIdentity();

  using OptimizerType = itk::RegularStepGradientDescentOptimizerv4<double>;
  auto optimizer = OptimizerType::New();
  optimizer->SetNumberOfIterations(200);
  optimizer->SetLearningRate(2.0);
  optimizer->SetMinimumStepLength(1e-4);

  using RegistrationType = itk::ImageRegistrationMethodv4<ImageType, ImageType, TransformType>;
  auto registration = RegistrationType::New();
  registration->SetFixedImage(fixed);
  registration->SetMovingImage(moving);
  registration->SetMetric(metric);
  registration->SetOptimizer(optimizer);
  registration->SetInitialTransform(transform);

  // Single resolution level (no shrink, no smoothing).
  RegistrationType::ShrinkFactorsArrayType   shrinkFactors(1);
  RegistrationType::SmoothingSigmasArrayType smoothingSigmas(1);
  shrinkFactors[0] = 1;
  smoothingSigmas[0] = 0;
  registration->SetNumberOfLevels(1);
  registration->SetShrinkFactorsPerLevel(shrinkFactors);
  registration->SetSmoothingSigmasPerLevel(smoothingSigmas);

  registration->Update();
  std::cout << "recovered parameters: " << registration->GetTransform()->GetParameters() << std::endl;

  // --- 3. resample the moving image with the recovered transform --------------------
  auto resample = itk::ResampleImageFilter<ImageType, ImageType>::New();
  resample->SetInput(moving);
  resample->SetTransform(registration->GetTransform());
  resample->SetUseReferenceImage(true);
  resample->SetReferenceImage(fixed);
  resample->SetDefaultPixelValue(0);
  itk::WriteImage(resample->GetOutput(), outputPath);

  return EXIT_SUCCESS;
}
